src/purple_vm/src/primitives/value.rs

```diff
@@ -149,6 +151,244 @@ impl VmValue {
+                    None => return Err(VmError::UnsafeCast) 
+                }
+            }
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    fn safe_f64_to_i64(operand: VmValue) -> Result<VmValue, VmError> {
+        match operand {
+            VmValue::F64(val) => {
+                if val.fract() != 0.0 {
+                    return Err(VmError::UnsafeCast);
+                }
+
+                match val.to_i64() {
+                    Some(res) => Ok(VmValue::I64(res)),
+                    None => return Err(VmError::UnsafeCast) 
+                }
+            }
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    fn safe_i32_to_f32(operand: VmValue) -> Result<VmValue, VmError> {
+        match operand {
+            VmValue::I32(val) => {
+                match val.to_f32() {
+                    Some(res) => Ok(VmValue::F32(res)),
+                    None => return Err(VmError::UnsafeCast)
+                }
+            }
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    fn safe_i64_to_f32(operand: VmValue) -> Result<VmValue, VmError> {
+        match operand {
+            VmValue::I64(val) => {
+                match val.to_f32() {
+                    Some(res) => Ok(VmValue::F32(res)),
+                    None => return Err(VmError::UnsafeCast)
+                }
+            }
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    // TODO: Check + test if value fits + check .fract() for equality if ToPrimitive doesn't handle
+    fn safe_f64_to_f32(operand: VmValue) -> Result<VmValue, VmError> {
+        match operand {
+            VmValue::F64(val) => {
+                match val.to_f32() {
+                    Some(res) => Ok(VmValue::F32(res)),
+                    None => return Err(VmError::UnsafeCast)
+                }
+            }
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    fn safe_i32_to_f64(operand: VmValue) -> Result<VmValue, VmError> {
+        match operand {
+            VmValue::I32(val) => {
+                match val.to_f64() {
+                    Some(res) => Ok(VmValue::F64(res)),
+                    None => return Err(VmError::UnsafeCast)
+                }
+            }
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    fn safe_i64_to_f64(operand: VmValue) -> Result<VmValue, VmError> {
+        match operand {
+            VmValue::I64(val) => {
+                match val.to_f64() {
+                    Some(res) => Ok(VmValue::F64(res)),
+                    None => return Err(VmError::UnsafeCast)
+                }
+            }
+            _ => return Err(VmError::InvalidOperand)
+        }
+    }
+
+    // TODO: Extend for arrays
+    pub fn i32_wrapi64 (&self) -> Result<VmValue, VmError> {
+        VmValue::safe_i64_to_i32(*self)
+    }
+
+    pub fn i32trunc_f32(&self) -> Result<VmValue, VmError> {
```

Comment:
The naming here seems to be flipped. The method name is `i32trunc_f32` when it should be `f32trunc_i32`. This is not intuitive and can result in serious programming errors.